interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,123 @@
+# file
+================
+
+## Overview
+
+The `file` module provides a platform-independent API for asynchronous file operations.
+
+-`file_create`: returns a file handle for the given file name.
+-`file_destroy`: closes the given file handle.
+-`file_write_async`: enqueues an asynchronous write request for a file at a given position.
+-`file_read_async`: enqueues an asynchronous read request for a file at a given position and size.
+-`file_extend_filesize`: expands the given file to be of desired size.
+
+## Exposed API
+
+```c
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+## file_create
+
+```c
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create,FILE_REPORT_FAULT, user_report_fault_callback, void*, user_report_fault_context, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name);
+```
+
+`file_create` opens an existing file by the name of `full_file_name` or creates a file of `desired_file_size` if it doesn't exist and returns the file handle.
+
+
+**SRS_FILE_43_033: [** If `execution_engine` is `NULL`, `file_create` shall fail and return `NULL`. **]**
+
+**SRS_FILE_43_002: [** If `full_file_name` is `NULL` then `file_create` shall fail and return `NULL`. **]**

```

Comment:
Thinking some more about the other requirements, I'm tempted to say this should go in the OS specific MD for the same reason I said that for "shall allocate a context". These requirements should be tested by unit tests and unit tests are OS specific.

Probably a good rule of thumb to put mostly requirements that can be tested by int tests in the interface MD, and the ones that need unit tests in the OS MD's.

Reply:
Also probably nice to say it shall fail if the string is an empty string. Separate requirement.

---
In reply to: [449272964](https://github.com/Azure/azure-c-pal/pull/7#discussion_r449272964) [](ancestors = 449272964)

Reply:
Er, I guess a better reason in this case for moving to the OS part is because you would write the unit tests for these requirements and we want to make sure it's written in both win32/Linux tests, so traceability can catch it by moving it to the OS specific specs.

---
In reply to: [449273161](https://github.com/Azure/azure-c-pal/pull/7#discussion_r449273161) [](ancestors = 449273161,449272964)